src/Stratis.Bitcoin/Connection/ConnectionManager.cs

```diff
@@ -264,13 +268,69 @@ public void Dispose()
             foreach (NetworkPeerServer server in this.Servers)
                 server.Dispose();
 
+            this.connectedPeersQueue.Dispose();
+
             this.networkPeerDisposer.Dispose();
         }
 
         /// <inheritdoc />
         public void AddConnectedPeer(INetworkPeer peer)
         {
             this.connectedPeers.Add(peer);
+            this.connectedPeersQueue.Enqueue(peer);
+        }
+
+        private Task OnPeerAdded(INetworkPeer peer, CancellationToken cancellationToken)
+        {
+            // Code in this method is a quick and dirty fix for the race condition described here: https://github.com/stratisproject/StratisBitcoinFullNode/issues/2864
+            // TODO race condition should be eliminated instead of fixing its consequences.
+            if (this.ShouldDisconnect(peer))
+                peer.Disconnect("Peer from the same network group.");
+
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Determines if the peer should be disconnected.
+        /// Peer should be disconnected in case it's IP is from the same group in which any other peer
+        /// is and the peer wasn't added using -connect or -addNode command line arguments.
+        /// </summary>
+        private bool ShouldDisconnect(INetworkPeer peer)
+        {
+            bool isAddNodeOrConnect = false;
+
+            foreach (IPEndPoint addNodeEndPoint in this.ConnectionSettings.AddNode.Union(this.ConnectionSettings.Connect))
+            {
+                if (peer.PeerEndPoint.Address.Equals(addNodeEndPoint.Address))
+                {
+                    isAddNodeOrConnect = true;
+                    break;
+                }
+            }
+
+            if (isAddNodeOrConnect)
+            {
+                this.logger.LogTrace("(-)[ADD_NODE_OR_CONNECT]:false");
+                return false;
+            }
+
+            byte[] peerGroup = peer.PeerEndPoint.MapToIpv6().Address.GetGroup();
```

Comment:
Dont we have a method to check same group already (the same method we check on connect)?

Reply:
No. There is a method in peer connector - `PeerIsPartOfExistingGroup` but it's private and should be moved somewhere else and it's interface should change in order to make it good to be used for other components